File: codetraverse/adapters/haskell_adapter.py
from tqdm import tqdm
from collections import defaultdict
import re

def adapt_haskell_components(raw_components):
    nodes = []
    edges = []
    node_ids = set()

    all_comps_by_module = defaultdict(list)
    main_modules_by_file = {
        comp['file_path']: comp['name']
        for comp in raw_components if comp.get("kind") == "module_header"
    }

    for comp in raw_components:
        name = comp.get("name")
        file_path = comp.get("file_path")
        if not (name and file_path): continue
        
        comp_module = comp.get("module", main_modules_by_file.get(file_path))
        if not comp_module: continue
            
        comp["module"] = comp_module
        all_comps_by_module[comp_module].append(comp)

    comps_by_file = defaultdict(list)
    for comp in raw_components:
        if comp.get("file_path"):
            comps_by_file[comp.get("file_path")].append(comp)

    for file_path, file_comps in comps_by_file.items():
        import_alias_map = {
            imp['alias']: imp['module']
            for imp in file_comps if imp.get("kind") == "import" and imp.get("alias")
        }
        
        for comp in file_comps:
            kind, name, comp_module = comp.get("kind"), comp.get("name"), comp.get("module")
            if not (kind and name and comp_module): continue
            
            source_id = f"{comp_module}::{name}"
            if source_id not in node_ids:
                node_category = kind
                if kind == "module_header": node_category = "module"
                if kind == "class": node_category = "typeclass"
                
                nodes.append({
                    "id": source_id, "category": node_category, "name": name,
                    "file_path": comp.get("file_path", ""),
                    "location": {"start": comp.get("start_line"),"end": comp.get("end_line")}
                })
                node_ids.add(source_id)

            if kind == "module_header":
                for export in comp.get("exports", []):
                    if export in import_alias_map:
                        alias = export
                        actual_module_names = [imp['module'] for imp in file_comps if imp.get("alias") == alias]

                        for actual_module_name in actual_module_names:
                            for target_comp in all_comps_by_module.get(actual_module_name, []):
                                target_name = target_comp.get("name")
                                if not target_name: continue
                                
                                proxy_id = f"{comp_module}::{target_name}"
                                if proxy_id not in node_ids:
                                    nodes.append({
                                        "id": proxy_id, "category": "reexport", "name": target_name,
                                        "file_path": comp.get("file_path", ""), "location": {}
                                    })
                                    node_ids.add(proxy_id)
                                
                                edges.append({"from": source_id, "to": proxy_id, "relation": "exports"})
                                actual_id = f"{target_comp['module']}::{target_name}"
                                edges.append({"from": proxy_id, "to": actual_id, "relation": "reexport_of"})
                    else:
                        edges.append({"from": source_id, "to": f"{comp_module}::{export}", "relation": "exports"})

            elif kind == "function":
                for call in comp.get("function_calls", []):
                    call_base = call.get("base", call.get("name"))
                    if not call_base: continue
                    target_module = call.get('modules', [comp_module])[0]
                    edges.append({"from": source_id, "to": f"{target_module}::{call_base}", "relation": "calls"})

                for dep in comp.get("type_dependencies", []):
                    dep_mod, _, dep_name = dep.rpartition('.')
                    edges.append({"from": source_id, "to": f"{dep_mod or comp_module}::{dep_name}", "relation": "uses_type"})
            
            elif kind == "instance":
                class_name = comp["name"].split()[0]
                class_id = f"{comp_module}::{class_name}"
                edges.append({"from": source_id, "to": class_id, "relation": "implements"})

    all_node_ids_final = {n['id'] for n in nodes}
    for edge in edges:
        for endpoint in ("from", "to"):
            if edge[endpoint] not in all_node_ids_final:
                nodes.append({"id": edge[endpoint], "category": "external", "name": edge[endpoint].split("::")[-1], "file_path": "external"})
                all_node_ids_final.add(edge[endpoint])
    
    print(f"Adapted {len(nodes)} nodes and {len(edges)} edges")
    return {"nodes": nodes, "edges": edges}

File: codetraverse/adapters/haskell_adapter.ts
import { Component, FunctionCall } from "../types/types";

interface Node {
  id: string;
  category: string;
  name: string;
  file_path: string;
  location?: {
    start?: number;
    end?: number;
  };
}

interface Edge {
  from: string;
  to: string;
  relation: string;
}

export function adaptHaskellComponents(rawComponents: Component[]): {
  nodes: Node[];
  edges: Edge[];
} {
  const nodes: Node[] = [];
  const edges: Edge[] = [];
  const nodeIds = new Set<string>();

  const allCompsByModule: Record<string, Component[]> = {};
  const mainModulesByFile: Record<string, string> = {};

  for (const comp of rawComponents) {
    if (comp.kind === "module_header") {
      mainModulesByFile[comp.filePath!] = comp.name;
    }
  }

  for (const comp of rawComponents) {
    const name = comp.name;
    const filePath = comp.filePath;
    if (!name || !filePath) continue;

    const compModule = comp.module || mainModulesByFile[filePath];
    if (!compModule) continue;

    comp.module = compModule;
    if (!allCompsByModule[compModule]) {
      allCompsByModule[compModule] = [];
    }
    allCompsByModule[compModule].push(comp);
  }

  const compsByFile: Record<string, Component[]> = {};
  for (const comp of rawComponents) {
    if (comp.filePath) {
      if (!compsByFile[comp.filePath]) {
        compsByFile[comp.filePath] = [];
      }
      compsByFile[comp.filePath].push(comp);
    }
  }

  for (const filePath in compsByFile) {
    const fileComps = compsByFile[filePath];
    const importAliasMap: Record<string, string> = {};
    for (const imp of fileComps) {
      if (imp.kind === "import" && imp.alias) {
        importAliasMap[imp.alias] = imp.module!;
      }
    }

    for (const comp of fileComps) {
      const { kind, name, module: compModule } = comp;
      if (!kind || !name || !compModule) continue;

      const sourceId = `${compModule}::${name}`;
      if (!nodeIds.has(sourceId)) {
        let nodeCategory = kind;
        if (kind === "module_header") nodeCategory = "module";
        if (kind === "class") nodeCategory = "typeclass";

        nodes.push({
          id: sourceId,
          category: nodeCategory,
          name: name,
          file_path: comp.filePath || "",
          location: { start: comp.startLine, end: comp.endLine },
        });
        nodeIds.add(sourceId);
      }

      if (kind === "module_header") {
        for (const exportName of comp.exports || []) {
          if (importAliasMap[exportName]) {
            const alias = exportName;
            const actualModuleNames = fileComps
              .filter((imp) => imp.alias === alias)
              .map((imp) => imp.module);

            for (const actualModuleName of actualModuleNames) {
              for (const targetComp of allCompsByModule[actualModuleName!] ||
                []) {
                const targetName = targetComp.name;
                if (!targetName) continue;

                const proxyId = `${compModule}::${targetName}`;
                if (!nodeIds.has(proxyId)) {
                  nodes.push({
                    id: proxyId,
                    category: "reexport",
                    name: targetName,
                    file_path: comp.filePath || "",
                    location: {},
                  });
                  nodeIds.add(proxyId);
                }

                edges.push({
                  from: sourceId,
                  to: proxyId,
                  relation: "exports",
                });
                const actualId = `${targetComp.module}::${targetName}`;
                edges.push({
                  from: proxyId,
                  to: actualId,
                  relation: "reexport_of",
                });
              }
            }
          } else {
            edges.push({
              from: sourceId,
              to: `${compModule}::${exportName}`,
              relation: "exports",
            });
          }
        }
      } else if (kind === "function") {
        for (const call of comp.functionCalls || []) {
          const callBase = call.base || call.name;
          if (!callBase) continue;
          const targetModule = (call.modules && call.modules[0]) || compModule;
          edges.push({
            from: sourceId,
            to: `${targetModule}::${callBase}`,
            relation: "calls",
          });
        }

        for (const dep of comp.typeDependencies || []) {
          const parts = dep.split(".");
          const depName = parts.pop()!;
          const depMod = parts.join(".");
          edges.push({
            from: sourceId,
            to: `${depMod || compModule}::${depName}`,
            relation: "uses_type",
          });
        }
      } else if (kind === "instance") {
        const className = comp.name.split(" ")[0];
        const classId = `${compModule}::${className}`;
        edges.push({ from: sourceId, to: classId, relation: "implements" });
      }
    }
  }

  const allNodeIdsFinal = new Set(nodes.map((n) => n.id));
  for (const edge of edges) {
    for (const endpoint of ["from", "to"]) {
      const edgePoint = edge[endpoint as keyof Edge];
      if (!allNodeIdsFinal.has(edgePoint)) {
        nodes.push({
          id: edgePoint,
          category: "external",
          name: edgePoint.split("::").pop()!,
          file_path: "external",
        });
        allNodeIdsFinal.add(edgePoint);
      }
    }
  }

  console.log(`Adapted ${nodes.length} nodes and ${edges.length} edges`);
  return { nodes, edges };
}
